File: kernel/backup-engine/src/test/kotlin/com/easydb/backup/RestorePolicyTest.kt
package com.easydb.backup

import com.easydb.common.DatabaseAdapter
import com.easydb.common.DatabaseCapabilities
import com.easydb.common.DbType
import io.mockk.every
import io.mockk.mockk
import org.junit.jupiter.api.Assertions.assertDoesNotThrow
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test

class RestorePolicyTest {
    private val damengAdapter = mockk<DatabaseAdapter>().also { adapter ->
        every { adapter.dbType() } returns DbType.DAMENG
        every { adapter.capabilities() } returns DatabaseCapabilities(
            supportsLogicalBackup = true,
            supportsLogicalRestore = true,
            supportsOverwriteRestore = false
        )
    }

    @Test
    fun `dameng rejects overwrite strategy`() {
        assertThrows(IllegalArgumentException::class.java) {
            RestorePolicy.validateOrThrow(
                config(strategy = "overwrite_existing"),
                manifest(),
                damengAdapter,
                targetExists = true
            )
        }
    }

    @Test
    fun `dameng rejects data only restore into new schema`() {
        assertThrows(IllegalArgumentException::class.java) {
            RestorePolicy.validateOrThrow(
                config(mode = "data_only"),
                manifest(),
                damengAdapter,
                targetExists = false
            )
        }
    }

    @Test
    fun `dameng rejects data only backup package`() {
        assertThrows(IllegalArgumentException::class.java) {
            RestorePolicy.validateOrThrow(
                config(),
                manifest(mode = "data_only"),
                damengAdapter,
                targetExists = false
            )
        }
    }

    @Test
    fun `restore to new rejects existing schema`() {
        assertThrows(IllegalArgumentException::class.java) {
            RestorePolicy.validateOrThrow(config(), manifest(), damengAdapter, targetExists = true)
        }
    }

    @Test
    fun `database type must match target adapter`() {
        assertThrows(IllegalArgumentException::class.java) {
            RestorePolicy.validateOrThrow(config(), manifest(dbType = "mysql"), damengAdapter, targetExists = false)
        }
    }

    @Test
    fun `selected table must exist in manifest`() {
        assertThrows(IllegalArgumentException::class.java) {
            RestorePolicy.validateOrThrow(
                config().copy(selectedTables = listOf("missing")),
                manifest(),
                damengAdapter,
                targetExists = false
            )
        }
    }

    @Test
    fun `mysql overwrite remains supported`() {
        val mysqlAdapter = mockk<DatabaseAdapter>().also { adapter ->
            every { adapter.dbType() } returns DbType.MYSQL
            every { adapter.capabilities() } returns DatabaseCapabilities(
                supportsLogicalBackup = true,
                supportsLogicalRestore = true,
                supportsOverwriteRestore = true
            )
        }

        assertDoesNotThrow {
            RestorePolicy.validateOrThrow(
                config(strategy = "overwrite_existing"),
                manifest(dbType = "mysql"),
                mysqlAdapter,
                targetExists = true
            )
        }
    }

    private fun config(mode: String = "restore_all", strategy: String = "restore_to_new") = RestoreConfig(
        targetConnectionId = "dm",
        backupFilePath = "/tmp/backup.edbkp",
        targetDatabase = "RESTORED_SCHEMA",
        mode = mode,
        strategy = strategy
    )

    private fun manifest(dbType: String = "dameng", mode: String = "full") = BackupManifest(
        formatVersion = 1,
        appVersion = "1.3.2",
        dbType = dbType,
        serverVersion = "8",
        database = "SOURCE_SCHEMA",
        mode = mode,
        startedAt = "2026-07-14T10:00:00",
        consistency = "snapshot",
        tables = emptyList(),
        objects = emptyList()
    )
}

File: apps/desktop-ui/src/utils/__tests__/restorePolicy.test.ts
import { describe, expect, it } from 'vitest'
import {
  isRestoreDatabaseTypeCompatible,
  isManifestModeRestorable,
  restoreModeOptions,
  restoreStrategyOptions,
  suggestedRestoreTarget,
} from '../restorePolicy'

describe('restore policy', () => {
  it('only offers restore-to-new for Dameng', () => {
    expect(restoreStrategyOptions('dameng')).toEqual([
      { label: '恢复到新 Schema（目标必须不存在）', value: 'restore_to_new' },
    ])
  })

  it('keeps overwrite restore available for MySQL', () => {
    expect(restoreStrategyOptions('mysql').map(option => option.value)).toEqual([
      'restore_to_new',
      'overwrite_existing',
    ])
  })

  it('requires the backup and target database types to match', () => {
    expect(isRestoreDatabaseTypeCompatible('DAMENG', 'dameng')).toBe(true)
    expect(isRestoreDatabaseTypeCompatible('mysql', 'dameng')).toBe(false)
  })

  it('does not reuse the selected existing Schema as the Dameng restore target', () => {
    expect(suggestedRestoreTarget('ENERGY', 'dameng', 'ENERGY')).toBe('ENERGY_restore')
    expect(suggestedRestoreTarget('shop', 'mysql', 'shop')).toBe('shop_restore')
    expect(suggestedRestoreTarget('shop', 'mysql', 'shop', 'overwrite_existing')).toBe('shop')
  })

  it('does not offer data-only restore into a new Dameng Schema', () => {
    expect(restoreModeOptions('dameng').map(option => option.value)).toEqual([
      'restore_all',
      'structure_only',
    ])
    expect(isManifestModeRestorable('data_only', 'dameng')).toBe(false)
    expect(isManifestModeRestorable('data_only', 'mysql')).toBe(true)
  })
})
